Add an option for a functionalized cp for TemperaturePressureFunctionFP

GiudGiud opened the pull request:
this is going to help us study the solution to https://github.com/idaholab/moose/issues/21245
It should also work a lot better for gases.

GiudGiud committed:
Add an option for variable cp for TP function-fp
Use a different definition to compute cp from cv
refs #21245

GiudGiud committed:
Add unit tests for the new variable cp option
refs #21245
Update regression test as cp expression changed

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2623244/), step Docs: sync website on 415decf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29641/site)

This comment will be updated on new commits.

GiudGiud commented on modules/fluid_properties/unit/include/TemperaturePressureFunctionFluidPropertiesTest.h:
```diff
@@ -55,7 +60,21 @@ class TemperaturePressureFunctionFluidPropertiesTest : public MooseObjectUnitTes
+
+    InputParameters uo_params2 =
+        _factory.getValidParams("TemperaturePressureFunctionFluidProperties");
+    // Set the three functions and the specific isobaric heat capacity
```
```suggestion
    // Set the three functions and the specific heat capacity
```

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C:
```diff
@@ -238,85 +355,150 @@ TemperaturePressureFunctionFluidProperties::h_from_p_T(
   Real rho, drho_dp, drho_dT;
   rho_from_p_T(pressure, temperature, rho, drho_dp, drho_dT);
-  return -drho_dT / rho;
+  return -drho_dp / rho;
```
that's a bug fix

joshuahansel replied:
I think it was right before.

GiudGiud replied:
https://en.wikipedia.org/wiki/Compressibility

ugh. we use beta for what they call alpha
beta in our dosctring is "beta volumetric thermal expansion coefficient"

joshuahansel replied:
Yeah that's pretty confusing. I see beta used for this coefficient sometimes, but alpha seems more common. At least the docstring appears correct and gives the units [1/K].

GiudGiud replied:
I ll change the notation in a follow up PR

GiudGiud commented on modules/fluid_properties/unit/src/TemperaturePressureFunctionFluidPropertiesTest.C:
```diff
@@ -98,16 +101,105 @@ TEST_F(TemperaturePressureFunctionFluidPropertiesTest, properties)
+{
+  const Real tol = REL_TOL_CONSISTENCY;
+  // Simpson integration only does 1e-5
+  const Real simpson_tol = 2e5 * tol;
```
this is rough but yeah 3/8 simpson is there in accuracy. Maybe we should have a tunable method with a n-point quadrature instead.

joshuahansel replied:
Also here when we deviate from one of our typical tolerances, I think it's preferred to not express it as a product but just write directly

GiudGiud replied:
ok I reworked that. Simpson was doing a good job but with a parameter for the integration we can now do better than this. p_from_v_e is still challenging tolerance-wise but I dont think it's incorrect

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2623414/), step Generate coverage on ff25cfe wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>44e87a</th>
      <th colspan=4>#29641 ff25cf</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.19%</td>
      <td>85.29%</td>
      <td>+0.10%</td>
      <td>93.96%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7777</td>
      <td>7873</td>
      <td>+96</td>
      <td>140</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1352</td>
      <td>1358</td>
      <td>+6</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29641/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29641/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29641/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
+\begin{aligned}
+  x = T \\
+  y = P \\
+  cp = cp^{user}(time=0, (x=T, y=P, z=0))
```
Use t for time; define if necessary

GiudGiud replied:
done

joshuahansel commented on modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C:
```diff
@@ -23,8 +23,13 @@ TemperaturePressureFunctionFluidProperties::validParams()
+      "cp", "Isobaric specific heat function of temperature and pressure [J/(kg-K)]");
+  params.addRangeCheckedParam<Real>(
+      "cv", 0, "cv >= 0", "Constant isochoric specific heat [J/(kg-K)]");
+  params.addParam<Real>("e_ref", 0, "Specific energy at the reference temperature");
+  params.addParam<Real>("T_ref", 0, "Reference temperature for the specific energy");
```
```suggestion
  params.addParam<Real>("e_ref", 0, "Specific internal energy at the reference temperature");
  params.addParam<Real>("T_ref", 0, "Reference temperature for the specific internal energy");
```

joshuahansel commented on modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C:
```diff
@@ -33,8 +38,15 @@ TemperaturePressureFunctionFluidProperties::validParams()
+    _T_ref(getParam<Real>("T_ref"))
 {
+  if (isParamValid("cp") && _cv_is_constant)
+    paramError("cp", "Can only set a function for cp if a constant is not specified for cv.");
```
```suggestion
    paramError("cp", "The parameter 'cp' may only be specified if 'cv' is unspecified or is zero.");
```

joshuahansel commented on modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C:
```diff
@@ -238,85 +355,150 @@ TemperaturePressureFunctionFluidProperties::h_from_p_T(
+        3. / 8. * h *
+        (cv_from_p_T(pressure, _T_ref) + 3 * cv_from_p_T(pressure, _T_ref + h) +
+         3 * cv_from_p_T(pressure, _T_ref + 2 * h) + cv_from_p_T(pressure, temperature));
+    return _e_ref + integral;
```
I think you suggested this in another comment, but maybe a parameter with a numerical integral interval would be best.

GiudGiud replied:
done

joshuahansel replied:
What are your thoughts on an external Simpson's integration method (my second comment)?

joshuahansel replied:
Ok, not a required change.

GiudGiud commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
 \end{equation}
 
-with $T$ the temperature, $P$ the pressure, $\rho$ the density, $\mu$ the dynamic viscosity, $k$ the thermal conductivity, $c_v$ the specific isochoric heat capacity, $e$ the specific energy and the $user$ exponent indicating a user-passed parameter.
+with $T$ the temperature, $P$ the pressure, $\rho$ the density, $\mu$ the dynamic viscosity, $k$ the thermal conductivity, $c_v$ the specific isochoric heat capacity, $e$ the specific energy, $\alpha$ the coefficient of thermal expansion, $\beta_T$ the isothermal
```
```suggestion
with $T$ the temperature, $P$ the pressure, $\rho$ the density, $\mu$ the dynamic viscosity, $k$ the thermal conductivity, $c_v$ the specific isochoric heat capacity, $e$ the specific internal energy, $T_{ref}$ a reference temperature at which the specific internal energy is equal to a reference energy $e_{ref}$, $\alpha$ the coefficient of thermal expansion, $\beta_T$ the isothermal
```

GiudGiud commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
+\begin{aligned}
+  x = T \\
+  y = P \\
+  cp = cp^{user}(time=0, (x=T, y=P, z=0))
```
```suggestion
  cp = cp^{user}(t=0, (x=T, y=P, z=0))
```

GiudGiud commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
   \rho = \rho^{user}(time=0, (x=T, y=P, z=0)) \\
   \mu = \mu^{user}(time=0, (x=T, y=P, z=0)) \\
   k = k^{user}(time=0, (x=T, y=P, z=0)) \\
```
```suggestion
  \rho = \rho^{user}(t=0, (x=T, y=P, z=0)) \\
  \mu = \mu^{user}(t=0, (x=T, y=P, z=0)) \\
  k = k^{user}(t=0, (x=T, y=P, z=0)) \\
```

GiudGiud commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
+\end{aligned}
+\end{equation}
+
+There are two options for specific heat. Either the user sets a constant specific isochoric heat capacity
```
```suggestion
Both the time (`t`) and Z-axis dimension are not used here. A fluid property made to depend on time will
not be properly updated by this `FluidProperties` object.
There are two options for specific heat. Either the user sets a constant specific isochoric heat capacity
```

GiudGiud committed:
Address Josh's review

- better doco
- fix beta (different notation than wikipedia) and cp-cv conversions
- use a parameterized integral for the e(T) = int(cv(T)) approximation
- adapt documentation
- adapt unit test to new values from numerical integration
- clean up tolerances in unit test

+
document remaining approximation in computation of e

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -12,12 +12,31 @@ functions. The following relations hold true:
+\begin{aligned}
   c_v = c_v^{user} \\
-  e = c_v * T
+  e = e_{ref} + c_v * (T - T_{ref})
```
done

GiudGiud committed:
Address FP unit test failures on M1 Mac

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -9,15 +9,36 @@ functions. The following relations hold true:
+  x = T \\
+  y = P \\
+  cp = cp^{user}(t=0, (x=T, y=P, z=0))
+  cv = cp - \dfrac{\alpha ^ 2 T}{\rho \beta_T}
```
```suggestion
  cv = cp - \dfrac{\alpha^2 T}{\rho \beta_T}
```

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/TemperaturePressureFunctionFluidProperties.md:
```diff
@@ -9,15 +9,36 @@ functions. The following relations hold true:
 
-with $T$ the temperature, $P$ the pressure, $\rho$ the density, $\mu$ the dynamic viscosity, $k$ the thermal conductivity, $c_v$ the specific isochoric heat capacity, $e$ the specific energy and the $user$ exponent indicating a user-passed parameter.
+with $T$ the temperature, $P$ the pressure, $\rho$ the density, $\mu$ the dynamic viscosity, $k$ the thermal conductivity, $c_v$ the specific isochoric heat capacity, $e$ the specific internal energy, $T_{ref}$ a reference temperature at which the specific internal energy is equal to a reference energy $e_{ref}$, $\alpha$ the coefficient of thermal expansion, $\beta_T$ the isothermal
+compressibility, and the $user$ exponent indicating a user-passed parameter.
```
I recommend some one-letter notation for the superscript. "alpha^user" looks like notation for a user-defined alpha coefficient.

joshuahansel commented on modules/fluid_properties/src/fluidproperties/TemperaturePressureFunctionFluidProperties.C:
```diff
@@ -23,8 +23,17 @@ TemperaturePressureFunctionFluidProperties::validParams()
+  params.addParam<Real>("e_ref", 0, "Specific internal energy at the reference temperature");
+  params.addParam<Real>("T_ref", 0, "Reference temperature for the specific internal energy");
+  params.addParam<unsigned int>(
+      "n_integration_intervals",
```
I think `dT_integration_intervals` would be a better parameter so that it's the same "resolution" for all temperatures.

GiudGiud committed:
Address Josh's review
- doco improvement
- add T to integration param name

joshuahansel commented:
I thought I had made a comment somewhere, but I can't find it: for the `n_intervals` thing for the time integration, I suggested using a specified interval size instead of number of intervals, for a more consistent accuracy

GiudGiud commented:
will do

GiudGiud committed:
Switch to an interval size based precision parameter for cv(T) integration to compute e

GiudGiud commented:
ok this makes more sense than what I understood. Added

GiudGiud commented:
Thanks for the review!
Test failures are unrelated
